### ISSUE TITLE
Incorrect nullability for `default(T)`

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#29849]
We should warn for a possible `null` assignment to `F3` and `F4`.

```C#
        [Fact]
        public void UnconstrainedGenericType()
        {
            var source =
@"internal class C<T> where T : new()
{
    internal T F1;
    internal T F2 = new T();
    internal T F3 = default; // missing warning
    internal T F4 = default(T); // missing warning
}";
            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
            comp.VerifyDiagnostics(
                // (1,16): warning CS8618: Non-nullable field 'F1' is uninitialized.
                // internal class C<T> where T : new()
                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F1").WithLocation(1, 16),
                // (3,16): warning CS0649: Field 'C<T>.F1' is never assigned to, and will always have its default value 
                //     internal T F1;
                Diagnostic(ErrorCode.WRN_UnassignedInternalField, "F1").WithArguments("C<T>.F1", "").WithLocation(3, 16)
                );
        }
```

**Comment by jcouv** [comment]
This was fixed by Aleksey's PR on type parameters.
Closing

```C#

        [Fact]
        public void NullabilityOfTypeParameters_029()
        {
            var source = @"
class Outer
{
    void M0<T>(T x)
    {
        x = default;
    }
}
";

            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
                // (6,13): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x = default;
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "default").WithLocation(6, 13)
                );
        }

        [Fact]
        public void NullabilityOfTypeParameters_030()
        {
            var source = @"
class Outer
{
    void M0<T>(T x)
    {
        x = default(T);
    }
}
";

            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
                // (6,13): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x = default(T);
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "default(T)").WithLocation(6, 13)
                );
        }

        [Fact]
        public void NullabilityOfTypeParameters_031()
        {
            var source = @"
class Outer
{
    void M0<T>(T x) where T : I1?
    {
        x = default;
    }
}

interface I1 {}
";

            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
                // (6,13): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x = default;
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "default").WithLocation(6, 13)
                );
        }

        [Fact]
        public void NullabilityOfTypeParameters_032()
        {
            var source = @"
class Outer
{
    void M0<T>(T x) where T : I1?
    {
        x = default(T);
    }
}

interface I1 {}
";

            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
                // (6,13): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x = default(T);
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "default(T)").WithLocation(6, 13)
                );
        }

        [Fact]
        public void NullabilityOfTypeParameters_033()
        {
            var source = @"
class Outer
{
    void M0<T>(T x) where T : Outer
    {
        x = default;
    }
}
";

            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
                // (6,13): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x = default;
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "default").WithLocation(6, 13)
                );
        }

        [Fact]
        public void NullabilityOfTypeParameters_034()
        {
            var source = @"
class Outer
{
    void M0<T>(T x) where T : Outer
    {
        x = default(T);
    }
}
";

            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
                // (6,13): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x = default(T);
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "default(T)").WithLocation(6, 13)
                );
        }
```
